README.md
# react-native-cross-geolocation

[![npm][npm-image]](https://www.npmjs.com/package/react-native-cross-geolocation)
[![License][license-image]](LICENSE)

React Native Geolocation complatible module that uses the new [Google Play services location API](https://developer.android.com/training/location/) on Android devices.

If this library has helped you, please support my work with a star or [buy me a coffee][kofi-url].

## IMPORTANT

This module was tested with React Native 0.59.0, but it should work smoothly with apps that use Gradle 4.6 or later.

\* For previous Gradle versions use react-native-cross-geolocation v1.0.6 or bellow.

## Setup

```bash
yarn add react-native-cross-geolocation
react-native link react-native-cross-geolocation
```

### Play Services Location Version

From v1.1.0, react-native-cross-geolocation supports the global variable `playServicesLocationVersion` to specify the version of 'com.google.android.gms:play-services-location' to use. It defaults to 16.0.0

### Configuration and Permissions

This section only applies to projects made with `react-native init` or to those made with Create React Native App which have since ejected. For more information about ejecting, please see the [guide](https://github.com/react-community/create-react-native-app/blob/master/EJECTING.md) on the Create React Native App repository.

#### iOS

You need to include the NSLocationWhenInUseUsageDescription key in Info.plist to enable geolocation when using the app. Geolocation is enabled by default when you create a project with react-native init.

In order to enable geolocation in the background, you need to include the 'NSLocationAlwaysUsageDescription' key in Info.plist and add location as a background mode in the 'Capabilities' tab in Xcode.

#### Android

To request access to location, add the following line to your app's AndroidManifest.xml:

```xml
<uses-permission android:name="android.permission.ACCESS_COARSE_LOCATION" />
<uses-permission android:name="android.permission.ACCESS_FINE_LOCATION" />
```

Android API >= 18 Positions will also contain a `mocked` boolean to indicate if position was created from a mock provider.
Android API >= 23 Permissions are handled automatically.

#### Methods

- [`setRNConfiguration`](#setrnconfiguration)
- [`requestAuthorization`](#requestauthorization)
- [`getCurrentPosition`](#getcurrentposition)
- [`watchPosition`](#watchposition)
- [`clearWatch`](#clearwatch)
- [`stopObserving`](#stopobserving)

## Reference

### Methods

#### `setRNConfiguration()`

```js
Geolocation.setRNConfiguration(config);
```

Sets configuration options that will be used in all location requests.

Parameters:

NAME | TYPE | REQUIRED | DESCRIPTION
---- | ---- | -------- | -----------
config | object | Yes | See below.

Supported options:

- `skipPermissionRequests` (boolean, iOS-only) - Defaults to `false`. If `true`, you must request permissions before using Geolocation APIs.
- `lowAccuracyMode` (number, Android-only) - Defaults to [LowAccuracyMode.BALANCED](#constants).
- `fastestInterval` (number, Android-only) - Defaults to 10000 (10 secs).
- `updateInterval` (number, Android-only) - Defaults to 5000 (5 secs).

#### `requestAuthorization()`

```js
Geolocation.requestAuthorization();
```

Request suitable Location permission based on the key configured on pList. If NSLocationAlwaysUsageDescription is set, it will request Always authorization, although if NSLocationWhenInUseUsageDescription is set, it will request InUse authorization.

#### `getCurrentPosition()`

```js
Geolocation.getCurrentPosition(geo_success, [geo_error], [geo_options]);
```

Invokes the success callback once with the latest location info.

Parameters:

NAME | TYPE | REQUIRED | DESCRIPTION
---- | ---- | -------- | -----------
geo_success | function | Yes | Invoked with latest location info.
geo_error | function | No | Invoked whenever an error is encountered.
geo_options | object | No | See below.

Supported options:

- `timeout` (ms) - Defaults to MAX_VALUE
- `maximumAge` (ms) - Defaults to INFINITY.
- `enableHighAccuracy` (bool) - On Android, if the location is cached this can return almost immediately, or it will request an update which might take a while.

#### `watchPosition()`

```js
Geolocation.watchPosition(success, [error], [options]);
```

Invokes the success callback whenever the location changes. Returns a `watchId` (number).

Parameters:

NAME | TYPE | REQUIRED | DESCRIPTION
---- | ---- | -------- | -----------
success | function | Yes | Invoked whenever the location changes.
error | function | No | Invoked whenever an error is encountered.
options | object | No | See below.

Supported options:

- `timeout` (ms) - Defaults to MAX_VALUE.
- `maximumAge` (ms) - Defaults to INFINITY.
- `enableHighAccuracy` (bool) - Defaults to `false`.
- `distanceFilter` (m) - Defaults to 100.
- `useSignificantChanges` (bool) (unused in Android).

#### `clearWatch()`

```js
Geolocation.clearWatch(watchID);
```

Parameters:

NAME | TYPE | REQUIRED | DESCRIPTION
---- | ---- | -------- | -----------
watchID | number | Yes | Id as returned by `watchPosition()`.

#### `stopObserving()`

```js
Geolocation.stopObserving();
```

Stops observing for device location changes. In addition, it removes all listeners previously registered.

Notice that this method has only effect if the `geolocation.watchPosition(successCallback, errorCallback)` method was previously invoked.

### Constants

rnCrossGeolocation exports a `LowAccuracyMode` object with values to be used in the `lowAccuracyMode` property of the parameter sent to [`setRNConfiguration`](#setrnconfiguration):

NAME | DETAILS
---- | -------
LowAccuracyMode.BALANCED<br>(102)<br>This is the default mode. | Request location precision to within a city block, which is an accuracy of approximately 100 meters.<br>This is considered a coarse level of accuracy, and is likely to consume less power. With this setting, the location services are likely to use WiFi and cell tower positioning. Note, however, that the choice of location provider depends on many other factors, such as which sources are available.
LowAccuracyMode.LOW_POWER<br>(104) | Request city-level precision, which is an accuracy of approximately 10 Km.<br>This is considered a coarse level of accuracy, and is likely to consume less power.
LowAccuracyMode.NO_POWER<br>(105) | Use this if you need negligible impact on power consumption, but want to receive location updates when available.<br>With this setting, your app does not trigger any location updates, but receives locations triggered by other apps.

_**NOTE:** These constants are only for Android, on iOS they are undefined._

## TODO

- [ ] Tests

## Support my Work

I'm a full-stack developer with more than 20 year of experience and I try to share most of my work for free and help others, but this takes a significant amount of time and effort so, if you like my work, please consider...

[<img src="https://amarcruz.github.io/images/kofi_blue.png" height="36" title="Support Me on Ko-fi" />][kofi-url]

Of course, feedback, PRs, and stars are also welcome 🙃

Thanks for your support!

[npm-image]:      https://img.shields.io/npm/v/react-native-cross-geolocation.svg
[license-image]:  https://img.shields.io/npm/l/express.svg
[kofi-url]:       https://ko-fi.com/C0C7LF7I


CHANGELOG.md
# react-native-cross-geolocation Changes

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/) and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## \[1.1.0] - 2018-03-27

### Added

- Missing Dependency on 'com.android.support:appcompat-v7'
- Markdownlint configuration.

### Changed

- Using Build Tools v28.0.3 with compileSdkVersion and targetSdkVersion 28
- Using Gradle plugin 3.2.1
- Using Java 1.8

### Removed

- Dependency on 'com.android.support:support-annotations'. This must fix conflict issues with versions.

## \[1.0.6] - 2018-10-10

### Fixed

- Error in getExtValue of build.gradle

## \[1.0.5] - 2018-10-09

### Fixed

- Error in build.gradle no getting the correct `googlePlayServicesVersion` variable.
- Minor fixes to the README.

## \[1.0.4] - 2018-09-09

### Changed

- Minimum Android SDK version from 21 to 16.
- Update README with a more clear note about the use of Gradle 4.4

### Fixed

- #2 : Error in build.gradle. Thanks to @mowbell for reporting this.

## \[1.0.3] - 2018-06-23

### Added

- Flow typings.

### Changed

- Updated README.
- The changelog follows the format on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).

### Fixed

- PR #1 Fixes `undefined` error by the use of Geolocation instead RNGeolocation. Thanks to @badrange

## \[1.0.2] - 2018-06-18

### Fixed

- `PositionError` builder

## \[1.0.1] - 2018-06-18

### Fixed

- Typings for `setRNConfiguration`

## [1.0.0] - 2018-06-18

- First public release


android/src/main/java/com/github/amarcruz/geolocation/LocationOptions.java
package com.github.amarcruz.geolocation;

import com.facebook.react.bridge.ReadableMap;
import com.google.android.gms.location.LocationRequest;

final class LocationOptions {
    private static final float DEFAULT_DISTANCE_FILTER = 100f;

    private static long fastestInterval = 5000;     // 5 secs
    private static long updateInterval = 10000;     // 10 secs
    private static int lowPriorityMode = LocationRequest.PRIORITY_BALANCED_POWER_ACCURACY;

    final long timeout;
    final double maximumAge;
    final float distanceFilter;
    final int priority;

    private LocationOptions(
            final long timeout,
            final double maximumAge,
            final boolean highAccuracy,
            final float distanceFilter
    ) {
        this.timeout = timeout;
        this.maximumAge = maximumAge;
        this.distanceFilter = distanceFilter;
        this.priority = highAccuracy ? LocationRequest.PRIORITY_HIGH_ACCURACY : lowPriorityMode;
    }

    static LocationOptions fromReactMap(final ReadableMap map) {
        long timeout = Long.MAX_VALUE;
        double maximumAge = Double.POSITIVE_INFINITY;
        boolean highAccuracy = false;
        float distanceFilter = DEFAULT_DISTANCE_FILTER;

        if (map != null) {
            // precision might be dropped on timeout (double -> int conversion), but that's OK
            if (map.hasKey("timeout")) {
                timeout = (long) map.getDouble("timeout");
            }
            if (map.hasKey("maximumAge")) {
                maximumAge = map.getDouble("maximumAge");
            }
            if (map.hasKey("enableHighAccuracy")) {
                highAccuracy = map.getBoolean("enableHighAccuracy");
            }
            if (map.hasKey("distanceFilter")) {
                distanceFilter = (float) map.getDouble("distanceFilter");
            }
        }

        return new LocationOptions(timeout, maximumAge, highAccuracy, distanceFilter);
    }

    static void setConfiguration (final ReadableMap map) {
        if (map != null) {
            if (map.hasKey("fastestInterval")) {
                fastestInterval = (long) map.getDouble("fastestInterval");
            }
            if (map.hasKey("updateInterval")) {
                updateInterval = (long) map.getDouble("updateInterval");
            }
            if (map.hasKey("lowPriorityMode")) {
                int mode = map.getInt("lowPriorityMode");
                if (mode != LocationRequest.PRIORITY_LOW_POWER && mode != LocationRequest.PRIORITY_NO_POWER) {
                    mode = LocationRequest.PRIORITY_BALANCED_POWER_ACCURACY;
                }
                lowPriorityMode = mode;
            }
        }
    }

    static long getFastestInterval () {
        return fastestInterval;
    }

    static long getUpdateInterval () {
        return updateInterval;
    }
}
